src/IW/App/Error.hs

```diff
@@ -0,0 +1,193 @@
+    ) where
+
+import Control.Monad.Except (MonadError)
+import Data.CaseInsensitive (foldedCase)
+import GHC.Stack (SrcLoc (SrcLoc, srcLocModule, srcLocStartLine))
+import Network.HTTP.Types.Header (HeaderName)
+import Servant.Server (err401, err404, err413, err417, err500, errBody)
+
+import qualified Control.Monad.Except as E (throwError)
+import qualified Servant.Server as Servant (ServerError)
+
+
+-- | Type alias for errors.
+type WithError m = (MonadError AppError m, HasCallStack)
+
+-- | Specialized version of 'E.throwError'
+throwError :: WithError m => AppErrorType -> m a
+throwError = E.throwError . AppError (toSourcePosition callStack)
+{-# INLINE throwError #-}
+
+newtype SourcePosition = SourcePosition Text
+    deriving newtype (Show, Eq)
+
+-- | Display 'CallStack' as 'SourcePosition' in a format: @Module.function#line_number@.
+toSourcePosition :: CallStack -> SourcePosition
+toSourcePosition cs = SourcePosition showCallStack
+  where
+    showCallStack :: Text
+    showCallStack = case getCallStack cs of
+        []                             -> "<unknown loc>"
+        [(name, loc)]                  -> showLoc name loc
+        (_, loc) : (callerName, _) : _ -> showLoc callerName loc
+
+    showLoc :: String -> SrcLoc -> Text
+    showLoc name SrcLoc{..} =
+        toText srcLocModule <> "." <> toText name <> "#" <> show srcLocStartLine
+
+{- | Exception wrapper around 'AppError'. Useful when you need to throw/catch
+'AppError' as 'Exception'.
+-}
+newtype AppException = AppException
+    { unAppException :: AppError
+    } deriving (Show)
+      deriving anyclass (Exception)
+
+-- | 'AppErrorType' with the corresponding 'CallStack'.
+data AppError = AppError
+    { appErrorCallStack :: !SourcePosition
+    , appErrorType      :: !AppErrorType
+    } deriving (Show, Eq)
+
+-- | App errors type.
+newtype AppErrorType = InternalError IError
+    deriving (Show, Eq)
+
+{- | The internal errors that can be thrown. These errors are meant to be
+handled within the application and cover exceptional circumstances/coding errors.
+-}
+data IError
+    {- | General not found. -}
+    = NotFound
+    {- | Some exceptional circumstance has happened stop execution and return.
+    Optional text to provide some context in server logs.
+    -}
+    | ServerError Text
+    {- | A required permission level was not met. Optional text to provide some context. -}
+    | NotAllowed Text
+    {- | Given inputs do not conform to the expected format or shape. Optional
+    text to provide some context in server logs.
+    -}
+    | Invalid Text
+    {- | Some header expected, but not present in header list.
+    -}
+    | MissingHeader HeaderName
+    {- | An authentication header that was required was provided but not in a
+    format that the server can understand
+    -}
+    | HeaderDecodeError Text
+    -- | Data base specific errors
+    | DbError Text
+    -- | Limits on the multi-request are overflowed.
+    | LimitError
```

Comment:
As discussed, we probably won't need `LimitError`. But there is no harm in having it :slightly_smiling_face:

Reply:
Oh yeah, I forgot about that. Got it.

Reply:
I'll just remove `LimitError` for now and bring it back later if needed.